### PR TITLE
Mean shift clustering

### PR DESCRIPTION
**Pull request by HurricaneTong**
I implement a basic mean shift algorithm.
I will appreciate it if anyone can give me some guides to make it accepted.

### CODE REVIEW

**Comment by rcurtin** [comment]
Hello there,

I'll take a look at this in the next couple of days and make some comments on the code.  We'll have to compare it against scikit's implementation for speed and accuracy, but that shouldn't be too hard.

**Comment by rcurtin** [comment]
Could you write some tests for the algorithm to verify that it works?  You could look at the existing tests in `src/mlpack/tests/` for inspiration.  Once we have some tests in place, I'll start digging into the Armadillo expressions to make sure they are as fast as possible.

**Comment by HurricaneTong** [comment]
I add a simple test using data in kmeans_test.cpp

File: src/mlpack/methods/mean_shift/mean_shift.hpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * @code
+ * extern arma::mat data; // Dataset we want to run Mean Shift on.
+ * arma::Col<size_t> assignments; // Cluster assignments.
+ * arma::mat centroids; // Cluster centroids.
+ *
+ * MeanShift<arma::mat, kernel::GaussianKernel> meanShift();
+ * meanShift.Cluster(dataset, assignments, centroids);
+ * @endcode
+ *
+ * @tparam KernelType the kernel to use.
```
Comment (rcurtin) [inline src/mlpack/methods/mean_shift/mean_shift.hpp]: Mean shift will not work with every type of kernel (i.e. it will not work with the linear kernel or the polynomial kernel or other kernels that are not shift-invariant and normalized).  We should make sure that the user knows that only certain kernels can be used.

File: src/mlpack/methods/mean_shift/mean_shift.hpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+   * @param duplicateThresh If distance of two centroids is less than it, one will be removed.
+   * @param maxIterations Maximum number of iterations allowed before giving up
+   * @param stopThresh If the 2-norm of the mean shift vector is less than stopThresh, 
+   *        iterations will terminate.
+   * @param kernel Optional KernelType object.
+   */
+  MeanShift(const double duplicateThresh = 1.0,
+            const size_t maxIterations = 1000,
+            const double stopThresh = 1e-3,
+            const KernelType kernel = KernelType());
```
Comment (rcurtin) [inline src/mlpack/methods/mean_shift/mean_shift.hpp]: I don't think that the duplicate threshold should have a default, because the value that should be used is completely dependent on the dataset.  I don't mind putting in some basic logic to estimate a threshold value if the user does not specify one, but a default of 1.0 is a bad idea, in my opinion.  (If you think I'm wrong, please feel free to provide a rebuttal.)

File: src/mlpack/methods/mean_shift/mean_shift_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    while (true) {
+      
+      // new centroid
+      arma::Col<double> newCentroid = arma::zeros(data.n_rows, 1);
+      
+      double sumWeight = 0;
+      for (size_t j = 0; j < data.n_cols; ++j) {
+        
+        // calc weight for each point
+        double weight = kernel.Evaluate(allCentroids.col(i), data.col(j));
```
Comment (rcurtin) [inline src/mlpack/methods/mean_shift/mean_shift_impl.hpp]: This is only correct for the Gaussian kernel, because d/dx(exp(-x/h)) = -1/h exp(-x/h).  For other kernels the expression g(x) might be substantially different (my notation comes from http://homepages.inf.ed.ac.uk/rbf/CVonline/LOCAL_COPIES/TUZEL1/MeanShift.pdf ).  I would suggest writing a standalone function that calculates the weight for a point, and then using a template specialization for the `GaussianKernel` and other kernels where you can just use the kernel evaluation.  In cases where the kernel is not radially symmetric, the case may be even more different (I have not derived that case myself).  Otherwise, you will probably have to require that each kernel has a `Gradient()` function and then you can call that when one of the template specializations is not being used.

File: src/mlpack/methods/mean_shift/mean_shift_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      
+      // calc the mean shift vector.
+      arma::Col<double> mhVector = newCentroid - allCentroids.col(i);
+      
+      // update the centroid.
+      allCentroids.col(i) = newCentroid;
+      
+      if (arma::norm(mhVector, 2) < stopThresh ||
+          completedIterations > maxIterations) {
+        break;
```
Comment (rcurtin) [inline src/mlpack/methods/mean_shift/mean_shift_impl.hpp]: This loop can be cleaner as a for loop instead of a `while(true)` where you manage `completedIterations` by hand.

File: src/mlpack/methods/mean_shift/mean_shift_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      }
+      
+    }
+    
+  }
+  
+  // remove duplicate centroids.
+  for (size_t i = 0; i < allCentroids.n_cols; ++i) {
```
Comment (rcurtin) [inline src/mlpack/methods/mean_shift/mean_shift_impl.hpp]: Why not check if there is a duplicate centroid once you finish finding the centroid for an individual point?  If you refactor the code to only consider one centroid at a time and only add it to the results if it is not duplicated in the existing list of results, I think you will get faster (and more memory-efficient) code.

File: src/mlpack/methods/mean_shift/mean_shift_main.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    << "no results will be saved." << std::endl;
+  }
+  
+  arma::mat dataset;
+  data::Load(inputFile, dataset, true); // Fatal upon failure.
+  arma::mat centroids;
+  arma::Col<size_t> assignments;
+  
+  kernel::GaussianKernel kernel;
+  kernel.Bandwidth(bandwidth);
```
Comment (rcurtin) [inline src/mlpack/methods/mean_shift/mean_shift_main.cpp]: Why not just use the constructor to set the bandwidth?

File: src/mlpack/tests/mean_shift_test.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  
+  // To ensure that this is the third class which we haven't seen yet.
+  BOOST_REQUIRE_NE(firstClass, thirdClass);
+  BOOST_REQUIRE_NE(secondClass, thirdClass);
+  
+  for (size_t i = 20; i < 30; i++)
+    BOOST_REQUIRE_EQUAL(assignments(i), thirdClass);
+  
+}
+
```
Comment (rcurtin) [inline src/mlpack/tests/mean_shift_test.cpp]: Once we merge this I will probably add at least one more test where I generate observations from a mixture of Gaussians (that are sufficiently far apart) and ensure that the mean shift algorithm will properly label each point.  You can do this now if you like, but I wanted to let you know that I'm planning to do that.  More tests, in general, is a good thing.  It may be worth adding some simple tests like "if I set the duplicate threshold to 0, do I get dataset.n_cols clusters?"; this is helpful to people who might be modifying the code later on and want to make sure it still functions the same way.

File: src/mlpack/methods/mean_shift/mean_shift.hpp (hunk 1 above)
Comment (stephentu) [inline src/mlpack/methods/mean_shift/mean_shift.hpp]: Maybe kernels can be tagged "translation_invariant" as a type trait? Then you can enforce this at compile time

File: src/mlpack/methods/mean_shift/mean_shift_main.cpp (hunk 6 above)
Comment (HurricaneTong) [inline src/mlpack/methods/mean_shift/mean_shift_main.cpp]: Because the bandwidth is a param of Gaussian kernel, not the param of MeanShift class. In the executable program, I don't provide param to choose kernel, the executable can only use Gaussian kernel, and there is a param to set the bandwidth of Gaussian kernel in the executable. But advanced user can use their own kernel by using class MeanShift.

File: src/mlpack/methods/mean_shift/mean_shift.hpp (hunk 1 above)
Comment (rcurtin) [inline src/mlpack/methods/mean_shift/mean_shift.hpp]: I thought about this a little more, and there's actually no need for type traits.  Translation invariant kernels can simply provide `double Evaluate(const double distance)`, so you might use this in the wild like:

```
// assume we have 'metric' and 'kernel'
const double k = kernel.Evaluate(metric.Evaluate(dataset.col(i), dataset.col(j)));
```

This also forces a radial basis kernel of some sort, so there's no need to worry about re-deriving mean shift for the non-radial kernel case.

File: src/mlpack/methods/mean_shift/mean_shift_main.cpp (hunk 6 above)
Comment (rcurtin) [inline src/mlpack/methods/mean_shift/mean_shift_main.cpp]: Yeah, I meant `kernel::GaussianKernel kernel(bandwidth);`. :)

File: src/mlpack/tests/mean_shift_test.cpp (hunk 7 above)
Comment (HurricaneTong) [inline src/mlpack/tests/mean_shift_test.cpp]: Two points may converge to one same centroid, so even duplicate thresh is set to 0, the number of  clusters may be still less than dataset.n_cols.

**Comment by HurricaneTong** [comment]
I add the metric and make some minor changes.
Now there is a function to provide estimation for duplicate thresh.
I do estimation like what scikit does.

**Comment by rcurtin** [comment]
Before I merge this, we still need to address different kernels that may have derivatives that are not the kernel function itself.  Like you've already done, we can assume that the kernel is shift-invariant by virtue of the fact that is has the function `Evaluate(double)`.  But what we really need to be using is a `Gradient()` function, so that will need to be added to all of the shift-invariant kernels that mlpack has.

**Comment by HurricaneTong** [comment]
In http://homepages.inf.ed.ac.uk/rbf/CVonline/LOCAL_COPIES/TUZEL1/MeanShift.pdf,
It seems kernel must be radially symmetric.
(using the notation in the paper above) What I use to calculate mean shift vector is the derivative of  k(s) rather than the derivative of K(x). 
For Gaussian kernel (u=1), dk(s)/ds = -0.5 * exp(-0.5 * s), dK(x)/dx = -x * exp(-0.5 * x). 
But for the function Gradient() in Gaussian Kernel class, I think it should implement dK(x)/dx.
It may be strange for Gradient() to implement dk(s)/ds.
I think it should have another name, but I really don't know what name is appropriate.

**Comment by HurricaneTong** [comment]
I make some changes to make my implementation in strict accordance with the paper above.
In the paper, only radially symmetric kernels are referred, 
so the template parameter metric won't make sense.
I think only Euclidean Distance is meaningful according to the deduction.
And in the past I use all points to calculate new centroid, 
now I only use points near the current centroid according to the paper.
I implement Gradient() in Gaussian Kernel as dk(s)/ds.
If it's appropriate, I will implement Gradient() in all radially symmetric kernels next.

**Comment by rcurtin** [comment]
I'm sorry for the slow response on this.  It took me some time to do some reading and some thinking.

First, some thoughts on the generalizability of mean shift clustering.  In the original paper by Fukunaga and Hostetler, the technique is derived for radially symmetric kernels, but could be re-derived for kernels satisfying a handful of conditions (the kernel is finite, the kernel decays to 0 as its inputs tend to infinity, the integral over the input space of the kernel is 1, the integral over the input space of the absolute value of the kernel is finite; see page 1 of the Fukunaga and Hostetler paper).  In my view, it is ideal to provide a general enough implementation for all kernels, but because we have not worked out the precise restrictions for non-radially-symmetric kernels and because the original formulation assumed radial symmetry, I think it is reasonable to assume a radially symmetric kernel.

For this paragraph, assume that we're using the Euclidean distance (I'll come back to this assumption later).  I'm also going to consider the bandwidth term as a function of the kernel k().  The derivation of the gradient of the density estimate is pretty simple (doing my best with the unicode here, and hoping I haven't made some stupid error):

∇f(x) = ∑ K'(∥ x - x_i ∥) d/dx(∥ x - x_i ∥).
      = ∑ K'(∥ x - x_i ∥) d/dx((∥ x - x_i ∥^2)^1/2)
      = ∑ K'(∥ x - x_i ∥) ((∥ x - x_i ∥^2)^(-1/2)) (x - x_i)

This is somewhat more complex than the nicer formulation in the paper you linked, which assumes that the kernel uses the squared distance ∥ x - x_i ∥^2 (for which the gradient is just 2(x - x_i)).  But we can't just plug in the squared Euclidean distance to every kernel, because the squared Euclidean distance does *not* satisfy the triangle inequality and is not a metric.  I haven't worked out the details, but this probably causes some stability issues with mean shift, and almost certainly would break the dual-tree implementation of mean shift, which will require the triangle inequality.

The Gaussian kernel is a nice exception, because it squares its input and thus ends up using the squared distance, yielding the nice expression which is easier to implement and calculate.

I said earlier I would revisit the assumption of the Euclidean distance.  Given the derivation above, it's pretty easy to plug in any d(x, x_i) in the place of ∥ x - x_i ∥ and then the gradient can be derived for that specific distance function.  I think it would be nice to generalize across possible distances, but this involves deriving the gradient for each of the distance metrics mlpack currently has implemented.  It would be nice if the mean shift implementation was sufficiently general to accept any distance metric (and then depend on its Gradient() function), but I don't see a reason to require it before accepting this contribution.

On the other hand, I do think it is absolutely necessary that the kernel type is parameterizable.  Implementing a Gradient() function in the various kernels which are shift-invariant is a good idea, but please don't implement dk(s)/ds, because this is not the actual gradient.  Instead, please implement the true gradient with respect to the input parameter (which is the distance between points), and use the long, ugly expression above which doesn't assume a squared distance.

Then, what we can do is use template specializations and SFINAE to catch those cases where the kernel uses the squared distance.  The GaussianKernel squares its input distance -- so it uses the squared distance.  You could add to the KernelTraits class and use SFINAE to specialize the gradient calculation for the GaussianKernel and other kernels that use the squared distance... something kind of like this:

```
// This is not necessarily the best design, just an example of SFINAE.
template<typename KernelType, typename MetricType> // MetricType if you want to make that generic too.
void MeanShift<KernelType, MetricType>::MeanShiftGradient(arma::mat& dataset,
                                                          typename boost::enable_if<!kernel::KernelTraits<KernelType>::UsesSquaredDistance>::type* junk = 0)
{
  // Slow, general implementation.
}

// This is not necessarily the best design, just an example of SFINAE.
template<typename KernelType, typename MetricType> // MetricType if you want to make that generic too.
void MeanShift<KernelType, MetricType>::MeanShiftGradient(arma::mat& dataset,
                                                          typename boost::enable_if<kernel::KernelTraits<KernelType>::UsesSquaredDistance>::type* junk = 0)
{
  // Super-fast implementation!
}
```

What do you think of this approach?  I think it's very important to provide generalized code where possible, and I think this is one of those instances.

**Comment by HurricaneTong** [comment]
Implement Gradient(const double distance) for shift-invariant kernels and
GradientForSquaredDistance(const double squaredDistance) for kernels whose UsesSquaredDistance is true to calculate weight of point faster.

**Comment by rcurtin** [comment]
I'm sorry, I think I may have over-thought this whole thing a bit.  Consider the more general derivation from before, for any kernel K and the Euclidean distance (I'm ignoring the constants on the front of f(x) for convenience.  You can rederive with the constants if you like):

∇f(x) = ∑ K'(∥ x - x_i ∥) d/dx(∥ x - x_i ∥).
= ∑ K'(∥ x - x_i ∥) d/dx((∥ x - x_i ∥^2)^1/2)
= ∑ K'(∥ x - x_i ∥) (1/2 (∥ x - x_i ∥^2)^(-1/2)) (x - x_i)
= ∑ K'(∥ x - x_i ∥) (1 / (2 ∥ x - x_i ∥^2)) (x - x_i)

Now plug in the Gaussian kernel:

K(∥ x - x_i ∥) = exp(-∥ x - x_i ∥^2 / b^2)

which gives by the chain rule

K'(∥ x - x_i ∥) = exp(-∥ x - x_i ∥^2 / b^2) * -b^-2 * 2 (∥ x - x_i ∥^2)

Note that last term, 2 (∥ x - x_i ∥^2).  When we plug that into ∇f(x) above, it cancels with the second term and we get the familiar

∇f(x) = K'(∥ x - x_i ∥) (x - x_i)

So there isn't actually a need for knowing whether or not the kernel uses a squared distance.  We can use this calculation always:

∇f(x) = ∑ K'(∥ x - x_i ∥) (1 / (2 ∥ x - x_i ∥^2)) (x - x_i)

and if the kernel _does_ use the squared distance, then there will exist the term 2 (∥ x - x_i ∥^2) in its gradient.  The `Gradient()` function should be inlined with the rest of the expression, so a smart compiler should be able to cancel out the 2 (∥ x - x_i ∥^2) and the (1 / (2 ∥ x - x_i ∥^2)).  And even if not, this is a few floating point calculations and not an unnecessary distance calculation (because you need to calculate ∥ x - x_i ∥ no matter what), so I don't think it will be incredibly computationally costly.  I'd be interested in comparing the runtime of the more general implementation with your Gaussian-specific implementation.  Ideally, the speed should be the same.

Does that make sense?  I'm sorry for the confusion.

**Comment by HurricaneTong** [comment]
Thanks for your detailed guidance.
I agree that my optimization for kernel which uses squared distance is unnecessary. 
A smart compiler should do it.
But I have a question about the general formulation.
I think it should be
∇f(x) = ∑ K'(∥ x - x_i ∥) d/dx(∥ x - x_i ∥).
= ∑ K'(∥ x - x_i ∥) d/dx((∥ x - x_i ∥^2)^1/2)
= ∑ K'(∥ x - x_i ∥) (1/2 (∥ x - x_i ∥^2)^(-1/2)) 2 * (x - x_i)
= ∑ K'(∥ x - x_i ∥) (1 / (∥ x - x_i ∥)) (x - x_i)
My implementation for general case accords with this formulation. 
I am going to remove the unnecessary kernel traits UsesSquaredDistance.

**Comment by rcurtin** [comment]
Oops, you're right, thanks for the fix.  My derivation of K'() for the Gaussian is wrong, too, it should be

K'(∥ x - x_i ∥) = exp(-∥ x - x_i ∥^2 / b^2) * -b^-2 * 2 (∥ x - x_i ∥)

The same reasoning still applies, though.

**Comment by rcurtin** [comment]
Okay, I finally had a chance to look through this more thoroughly (my schedule has cleared up a bit, and I can now focus on getting this done).  I went through and made several changes, but it was unclear to me how I might commit to your particular branch, so what I did instead is post my patches to http://www.ratml.org/misc/patches.zip.  You can unzip the patches, then 'git am *.patch' to apply them to your branch.  It's mostly formatting/style cleanups, but there are a few speedups there.

However, what kept me from saying "okay, we're good to go" is that the scikit implementation of mean shift is an order of magnitude faster or more.  I used the following test script for scikit:

```
import numpy
from sklearn.cluster import MeanShift, estimate_bandwidth

d = numpy.genfromtxt('/path/to/dataset.csv', delimiter=',')
bw = estimate_bandwidth(d, quantile=0.2, n_samples=500)

print(bw)

ms = MeanShift(bandwidth=bw, bin_seeding=True)
ms.fit(d)

print(ms.cluster_centers_)
print(len(numpy.unique(ms.labels_)))
```

And then I would use the printed bandwidth with the `mean_shift` program.  Could you please look into why the implementation for mlpack is so slow comparatively?

**Comment by HurricaneTong** [comment]
Thanks for your patches. I feel very guilty for letting you do these trivial work like modifying code style. But I can't apply 0002-Remove-spaces-remove-unnecessary-semicolons.patch, git always failed when applying it. So I just remove two unnecessary semicolons manually. 
I do tests on iris.csv. scikit costs about 30 ms and mlpack costs 200 ms. After using unsafe_col, mlpack costs 140ms. I think the major cause is range search.
In my implementation I use brute way to do range search. But scikit uses tree to do this.
I do following changes in mean_shift_.py line 131 in scikit 
            #i_nbrs = nbrs.radius_neighbors([my_mean], bandwidth,
            #                               return_distance=False)[0]
            i_nbrs = []
            for i in range(0, X.shape[0]) :
                if (extmath.norm(my_mean - X[i]) < bandwidth) :
                    i_nbrs.append(i)
With this brute approach, scikit costs 400ms.
I go through RangeSearch in mlpack but it seems I can only do search with pre-determined querySet.
How can I do search if I have a pre-determined referenceSet  but don't know the querySet ahead.

**Comment by rcurtin** [comment]
Okay, I finally got a chance to look into this.  I've refactored the `RangeSearch` class in 3673579 so that you can pass different query sets.  You can take a look at `range_search.hpp` to see what I've changed.  Can you give this a try and see if this speeds things up?

I'm planning on adapting the API of `NeighborSearch` and `FastMKS` too, so seeing if you have any problems here will be helpful for better refactoring. :)

**Comment by HurricaneTong** [comment]
I apply range searcher and add a function to generate seeds from the data set as initial centroids like what scikit does. I use all the points as initial centroids before.  Generated seeds is much fewer than all the points.
I used the following test script for scikit:
```
import numpy
import time
from sklearn.cluster import MeanShift, estimate_bandwidth

d = numpy.genfromtxt('iris.csv', delimiter=',')
bw = estimate_bandwidth(d, quantile=0.2, n_samples=500)

print(bw)

ms = MeanShift(bandwidth=bw, bin_seeding = True)
t1 = time.time()
ms.fit(d)
t2 = time.time()
print t2 - t1


print(ms.cluster_centers_)
print(len(numpy.unique(ms.labels_)))
```
the result is 0.038s
0.912643298082
0.0381479263306
[[ 6.28301887  2.88679245  4.90754717  1.7       ]
 [ 4.97391304  3.39130435  1.47391304  0.24130435]]
2
then I call MS program with 
-v -i iris.csv -o assignments.csv -C centroids.csv -r 0.912643298082
I get 
clustering: 0.048295s
MS program is much faster than before.
But my implementation is still slower than scikit because of the algorithm.
```
          distances[0][j] /= radius;
          double weight = kernel.Gradient(distances[0][j]) / distances[0][j];
          sumWeight += weight;
          newCentroid += weight * data.unsafe_col(neighbors[0][j]);
```
I use this to iterate and can use user-defined kernel.
But scikit just use the average 
```
my_mean = np.mean(points_within, axis=0)
```
I tried to use the same simple algorithm like scikit, then MLPACK and scikit cost the same time.

**Comment by rcurtin** [comment]
Do you think you can use a template specialization here to use the simple average when you can, but when you can't, use the general iterative procedure?  If you can do that, we can run some more benchmarks and commit it if the benchmarks look good. :)

**Comment by HurricaneTong** [comment]
I add simple mean approach then scikit and my implementation costs nearly the same on iris.csv.
Could you give some more benchmarks? I don't know where to find appropriate benchmarks.

**Comment by rcurtin** [comment]
Yeah, here are some you can try that I've thrown on a server:

http://www.ratml.org/misc/isolet.csv -- spoken letters of the alphabet; 617 dimensions, 7797 points
http://www.ratml.org/misc/birch3.txt -- BIRCH dataset, 2 dimensions, 100000 points
http://www.ratml.org/misc/corel.csv -- corel image features dataset; 32 dimensions, 37749 points
http://www.ratml.org/misc/covertype.csv -- forest covertype dataset; 55 dimensions, 581012 points

These datasets will probably provide a good picture of the speed of your implementation vs. scikit.

**Comment by HurricaneTong** [comment]
I tried corel.csv.
My implementation costs 44s
```
[0;32m[INFO ] [0mProgram timers:
[0;32m[INFO ] [0m  clustering: 44.613388s
[0;32m[INFO ] [0m  computing_neighbors: 0.114881s
[0;32m[INFO ] [0m  loading_data: 7.664288s
[0;32m[INFO ] [0m  range_search/computing_neighbors: 37.602321s
[0;32m[INFO ] [0m  range_search/tree_building: 1.618025s
[0;32m[INFO ] [0m  saving_data: 0.639427s
[0;32m[INFO ] [0m  total_time: 52.928002s
[0;32m[INFO ] [0m  tree_building: 1.622805s
```
While scikit costs 10.5s.
Result is the same.
I think the bottleneck is range_search.
I compare the range_search of mlpack with NearestNeighbors of scikit
I save all queried vectors when performing MS into a means.csv and run query.
```
  range::RangeSearch<> rangeSearcher(data, false, true);
  arma::mat tAllMean = allMean.t();
  tAllMean.save("means.csv", arma::csv_ascii);
  Timer::Start("search_test");
  for (size_t i = 0; i < allMean.n_cols; ++i) {
    rangeSearcher.Search(allMean.unsafe_col(i), validRadius,
                         neighbors, distances);

  }
  Timer::Stop("search_test");
  timeval t = Timer::Get("search_test");
  std::cout << t.tv_sec << std::endl;
```
it costs 38s
while in scikit,
```
d = numpy.genfromtxt('/Users/HurricaneTong/GitHub/mlpack/build_MS_nondebug/bin/Debug/corel.csv', delimiter=',')
bw = estimate_bandwidth(d, quantile=0.2, n_samples=500)


means = numpy.genfromtxt('/Users/HurricaneTong/GitHub/mlpack/build_MS_nondebug/bin/Debug/means.csv', delimiter=',')
nbrs = NearestNeighbors(radius=bw).fit(d)
t1 = time.time()
for i in range(0, means.shape[0]):
    nbrs.radius_neighbors([means[i,:]], bw, return_distance=True)
t2 = time.time()
print t2 - t1
```
it costs 3.4s

**Comment by rcurtin** [comment]
Are you sure you compiled mlpack without debugging symbols?  Here is what I get when compiling mlpack with `-DDEBUG=OFF` and `-DPROFILE=OFF`.  I used this test program for scikit:

```
#!/usr/bin/python
import numpy
from sklearn.cluster import MeanShift
from sklearn.cluster import estimate_bandwidth
import time

d = numpy.genfromtxt('/home/ryan/datasets/corel.csv', delimiter=',')
bw = estimate_bandwidth(d, quantile=0.2, n_samples=500)

print(bw)

ms = MeanShift(bandwidth=bw, bin_seeding = True)
t1 = time.time()
ms.fit(d)
t2 = time.time()

print t2 - t1

print(len(numpy.unique(ms.labels_)))
```

This gave me the following output:

```
0.430335887828
7.03606009483
1
```

So, bandwidth of 0.430336, it took 7.036 seconds, and we got 1 cluster as a result.  Then, I use your implementation for mlpack:

```
$ mean_shift -i ~/datasets/corel.csv -r 0.430335887828 -v -C centers.csv
[INFO ] Loading '/home/ryan/datasets/corel.csv' as CSV data.  Size is 32 x 37749.
[INFO ] Performing mean shift clustering...
[INFO ] 46511 node combinations were scored.
[INFO ] 37749 base cases were calculated.
[INFO ] Found 1 centroids.
[WARN ] No extension given with filename ''; type unknown.  Save failed.
[INFO ] Saving CSV data to 'centers.csv'.
[INFO ] 
[INFO ] Execution parameters:
[INFO ]   bandwidth: (Unknown data type - )
[INFO ]   centroid_file: centers.csv
[INFO ]   help: false
[INFO ]   in_place: false
[INFO ]   info: ""
[INFO ]   inputFile: /home/ryan/datasets/corel.csv
[INFO ]   max_iterations: 1000
[INFO ]   output_file: ""
[INFO ]   radius: 0.430336
[INFO ]   verbose: true
[INFO ]   version: false
[INFO ] 
[INFO ] Program timers:
[INFO ]   clustering: 3.681358s
[INFO ]   computing_neighbors: 0.009845s
[INFO ]   loading_data: 0.459559s
[INFO ]   range_search/computing_neighbors: 2.075936s
[INFO ]   range_search/tree_building: 0.440392s
[INFO ]   saving_data: 0.000118s
[INFO ]   total_time: 4.143638s
[INFO ]   tree_building: 0.487500s
```

So, the mlpack implementation appears to be twice as fast as the scikit implementation.  (I'm using Python 2.7.9 with Debian's `python-sklearn` 0.15.2-3 package.)  I wouldn't be surprised if newer versions of scikit seem faster, but either way, the timings I'm getting are drastically different than you are, so maybe there is a configuration issue on your end?

With the covertype dataset and a bandwidth of 1524.6535, scikit takes 157.3325s while mlpack takes 42.159s.

**Comment by rcurtin** [comment]
Also, the warning `[WARN ] No extension given with filename ''; type unknown.  Save failed.` is caused by line 52 in `mean_shift_main.cpp`:

`const double bandwidth = CLI::GetParam<double>("bandwidth");`

That variable isn't used, and isn't documented as a `PARAM_DOUBLE()` macro.  My guess is that the line isn't necessary. :)

**Comment by HurricaneTong** [comment]
oh.... I just use -DDEBUG=OFF and don't set -DPROFILE=OFF. :)
I also find the WARN... I deleted the useless line but haven't pushed it to github.

**Comment by HurricaneTong** [comment]
I correct my configuration and get the same result with you and delete the useless line.
Thanks for pointing it out.

**Comment by rcurtin** [comment]
Okay, great.  I think this is ready to merge, then.  If you agree, I'll go ahead and merge it manually. :)

**Comment by HurricaneTong** [comment]
I'd be absolutely delighted if my implementation can be merged. Thanks for your help during the long time.
